### ISSUE TITLE
unprivileged podman build . fails EPERM on NFS, fedora29

### ISSUE THREAD

**Issue by fche** [containers/buildah#1387]
**Description**

**Steps to reproduce the issue:**
1. git clone https://github.com/openshift/prometheus foobar
2. cd foobar
3. podman build .
4. sudo podman build .

**Describe the results you received:**

```
STEP 1: FROM registry.svc.ci.openshift.org/openshift/release:golang-1.10 AS builder
Getting image source signatures
Copying blob a02a4930cb5d: 71.68 MiB / 71.68 MiB [==========================] 9s
Copying blob aad0e63dc7e2: 274.93 MiB / 274.93 MiB [========================] 9s
Copying config 6cb8ecdeb577: 3.41 KiB / 3.41 KiB [==========================] 0s
Writing manifest to image destination
Storing signatures
ERRO[0014] Error while applying layer: ApplyLayer exit status 1 stdout:  stderr: open /root/.bash_logout: permission denied 
error creating build container: Error committing the finished image: error adding layer with blob "sha256:a02a4930cb5d36f3290eb84f4bfa30668ef2e9fe3a1fb73ec015fc58b9958b17": ApplyLayer exit status 1 stdout:  stderr: open /root/.bash_logout: permission denied
```

**Describe the results you expected:**

success, as per the `sudo podman build .` run.

**Output of `rpm -q buildah` or `apt list buildah`:**

```
buildah-1.7-1.git873f001.fc29.x86_64
podman-1.0.0-1.git82e8011.fc29.x86_64

```

**Output of `buildah version`:**

```
Version:         1.7
Go Version:      go1.11.5
Image Spec:      1.0.0
Runtime Spec:    1.0.0
CNI Spec:        0.4.0
libcni Version:  
Git Commit:      
Built:           Wed Dec 31 19:00:00 1969
OS/Arch:         linux/amd64
```

**Output of `podman version` if reporting a `podman build` issue:**

```
Version:       1.0.0
Go Version:    go1.11.4
Git Commit:    "49780a1cf10d572edc4e1ea3b8a8429ce391d47d"
Built:         Mon Jan 14 15:38:17 2019
OS/Arch:       linux/amd64

```

**Output of `cat /etc/*release`:**

```
NAME=Fedora
VERSION="29 (Workstation Edition)"
etc.
```

**Output of `uname -a`:**

```
Linux HOSTNAME 4.19.4-300.fc29.x86_64 #1 SMP Fri Nov 23 13:03:11 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by rhatdan** [comment]
This looks like it thinks the root directory is where containers storage should be stored?

**Comment by giuseppe** [comment]
What is you `env` when running as rootless?

**Comment by fche** [comment]
There's lots in there, including personal stuff.  Which variables in specific do you need?

**Comment by rhatdan** [comment]
@fche Lets start by putting the machine in permissive mode to see if SELinux is blocking something.

**Comment by fche** [comment]
% getenforce
Permissive

sorry for making you cry. :-)

**Comment by giuseppe** [comment]
USER and XDG_* variables

**Comment by fche** [comment]
```
XDG_SEAT=seat0
XDG_SESSION_ID=1
XORG_RUN_AS_USER_OK=1
XDG_RUNTIME_DIR=/run/user/500
USER=fche
XDG_DATA_DIRS=/home/fche/.local/share/flatpak/exports/share:/var/lib/flatpak/exports/share:/usr/local/share:/usr/share
XDG_VTNR=1
```

**Comment by fche** [comment]
(I was not able to find usable strace-level diagnostics of what was going wrong, because `sudo strace -f podman ...` took it down a working pathway.)

**Comment by giuseppe** [comment]
can you also share your `~/.config/containers/libpod.conf` and `~/.config/containers/storage.conf`?

**Comment by fche** [comment]
```
/home/fche/.config/containers/libpod.conf:volume_path = "/home/fche/.local/share/containers/storage/volumes"
/home/fche/.config/containers/libpod.conf:image_default_transport = "docker://"
/home/fche/.config/containers/libpod.conf:runtime_path = ["/usr/bin/runc", "/usr/sbin/runc", "/usr/local/bin/runc", "/usr/local/sbin/runc", "/sbin/runc", "/bin/runc", "/usr/lib/cri-o-runc/sbin/runc"]
/home/fche/.config/containers/libpod.conf:conmon_path = ["/usr/libexec/podman/conmon", "/usr/libexec/crio/conmon", "/usr/local/lib/podman/conmon", "/usr/local/libexec/crio/conmon", "/usr/bin/conmon", "/usr/sbin/conmon", "/usr/lib/crio/bin/conmon"]
/home/fche/.config/containers/libpod.conf:conmon_env_vars = ["PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"]
/home/fche/.config/containers/libpod.conf:cgroup_manager = "cgroupfs"
/home/fche/.config/containers/libpod.conf:init_path = "/usr/libexec/podman/catatonit"
/home/fche/.config/containers/libpod.conf:static_dir = "/home/fche/.local/share/containers/storage/libpod"
/home/fche/.config/containers/libpod.conf:tmp_dir = "/run/user/500/libpod/tmp"
/home/fche/.config/containers/libpod.conf:max_log_size = -1
/home/fche/.config/containers/libpod.conf:no_pivot_root = false
/home/fche/.config/containers/libpod.conf:cni_config_dir = "/etc/cni/net.d/"
/home/fche/.config/containers/libpod.conf:cni_plugin_dir = ["/usr/libexec/cni", "/usr/lib/cni", "/usr/local/lib/cni", "/opt/cni/bin"]
/home/fche/.config/containers/libpod.conf:infra_image = "k8s.gcr.io/pause:3.1"
/home/fche/.config/containers/libpod.conf:infra_command = "/pause"
/home/fche/.config/containers/libpod.conf:enable_port_reservation = true
/home/fche/.config/containers/libpod.conf:label = true
/home/fche/.config/containers/storage.conf:[storage]
/home/fche/.config/containers/storage.conf:  driver = "overlay"
/home/fche/.config/containers/storage.conf:  runroot = "/run/user/500"
/home/fche/.config/containers/storage.conf:  graphroot = "/home/fche/.local/share/containers/storage"
/home/fche/.config/containers/storage.conf:  [storage.options]
/home/fche/.config/containers/storage.conf:    mount_program = "/usr/bin/fuse-overlayfs"
````

**Comment by rhatdan** [comment]
@giuseppe Still working on this?

**Comment by fche** [comment]
Another data point that may be relevant: this unprivileged userid has an nfs-mounted /home.
On a lark, I created another such user, but with a `/var/something` home, and that worked.

**Comment by rhatdan** [comment]
NFS Might be being blocked on norootsquash.

**Comment by fche** [comment]
This one is exported with no_root_squash.  Maybe overlayfs & nfs don't get along within the kernel?

**Comment by rhatdan** [comment]
@rhvgoyal Does overlay work on a nfs share?

@fche Does buildah unshare work?

**Comment by fche** [comment]
```
% buildah unshare podman pull registry.svc.ci.openshift.org/openshift/origin-v4.0:base
could not get runtime: remove /home/fche/.local/share/containers/storage/vfs/dir/91e4cddb1d38553ddef8329e17e0082b23575b4f5cc5a38ea0498c85b37d1e66/temp-storage-extract331098753: permission denied
ERRO[0002] exit status 125                              
ERRO[0002] exit status 125     

% sudo rm -rf /home/fche/.local/share/containers/        

%  buildah unshare podman pull registry.svc.ci.openshift.org/openshift/origin-v4.0:base
Trying to pull registry.svc.ci.openshift.org/openshift/origin-v4.0:base...Getting image source signatures
Copying blob 68cddb23acfe: 72.31 MiB / 72.31 MiB [====================] 3s
Copying blob b1ae8487cc2f: 975 B / 975 B [============================] 3s
Copying blob ad0d4b936354: 472 B / 472 B [============================] 3s
Copying blob 511ef889c4a1: 7.41 MiB / 7.41 MiB [======================] 3s
Copying config 3d1578b789ba: 4.73 KiB / 4.73 KiB [====================] 0s
Writing manifest to image destination
Storing signatures
Failed
error pulling image "registry.svc.ci.openshift.org/openshift/origin-v4.0:base": unable to pull registry.svc.ci.openshift.org/openshift/origin-v4.0:base: unable to pull image: Error committing the finished image: error adding layer with blob "sha256:68cddb23acfeddaee12b95b560b510d2ce2643a3c6a892d9df10da63a3089e78": Error processing tar file(exit status 1): lchown /run/systemd/netif: operation not permitted
ERRO[0011] exit status 125                              
ERRO[0011] exit status 125                        
```
```

**Comment by rhatdan** [comment]
I just wanted to see if 
buildah unshare /bin/sh
worked?

And then could you change uids and create content with different UIDs.

**Comment by fche** [comment]
```
 % buildah unshare /bin/sh 
sh-4.4# id
uid=0(root) gid=0(root) groups=0(root),65534(nfsnobody) context=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
sh-4.4# pwd
/home/fche
sh-4.4# touch foo
```
in another shell:
```
ls -al /home/fche/foo
-rw-r--r--. 1 fche users 0 Mar  8 12:06 /home/fche/foo
```

so that looks like it works.

**Comment by rhatdan** [comment]
@fche While in the unshare
su bin
touch /home/fche/bin_file

**Comment by fche** [comment]
```
% buildah unshare /bin/sh
sh-4.4# su bin
This account is currently not available.
sh-4.4# su -s /bin/sh bin
sh-4.4$ touch /home/fche/bin_file
touch: cannot touch '/home/fche/bin_file': Permission denied
sh-4.4$ 
```

**Comment by fche** [comment]
OTOH
```
% mkdir foobar
% chmod 777 foobar
# [in buildah unshare] touch foobar/file [works]
% [back outside] ls -al foobar
% ls -al /home/fche/foobar
total 76
drwxrwxrwx.   2 fche   users   4096 Mar  8 12:11 .
drwxr-x--x. 356 fche   users  65536 Mar  8 12:11 ..
-rw-r--r--.   1 624108 624108     0 Mar  8 12:11 bin_file
```

**Comment by rhatdan** [comment]
Is you example above inside of our out of the container?

@giuseppe This could be a fuse-overlay issue on top of NFS.

**Comment by fche** [comment]
The "back outside" part is outside.
The "in ... unshare" is inside.

**Comment by rhatdan** [comment]
Looks like the bin_file got created with a ramond UID.

[ in buildah unshare ] cat /proc/self/uid_map; id

**Comment by fche** [comment]
```
% buildah unshare /bin/sh
sh-4.4# cat /proc/self/uid_map; id
         0        500          1
         1     624108      65536
uid=0(root) gid=0(root) groups=0(root),65534(nfsnobody) context=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
sh-4.4# su -s /bin/sh bin
sh-4.4$ id
uid=1(bin) gid=1(bin) groups=1(bin) context=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
```

**Comment by rhatdan** [comment]
Ok so this is working correctly.

**Comment by rhatdan** [comment]
@giuseppe My best guess is this has something to do with Fuse-overlay on top of NFS.

**Comment by giuseppe** [comment]
please take a look at the nfs4 ACLs, it might be denied to the user the permission to change ownership for the files.  I went farther doing that, but got stuck on the ID mapping when using an ID>0 in the user namespace, since the user could not be mapped.

**Comment by giuseppe** [comment]
@dhowells is NFS supposed to work within a user namespace?

**Comment by rhatdan** [comment]
We have documented that User Namespace does not work on NFS directories.

**Comment by rhatdan** [comment]
@ashley-cui Could you add a comment on this in the troublshoot.md file, that rootless buildah does not work on nfs. You can probably grab the documentation from the libpod/rootless.md page.  Then we can close this issue.

**Comment by ashley-cui** [comment]
Is there a solution to this other than not using NFS and not using rootless?

**Comment by rhatdan** [comment]
Yes, if you setup container storage in a different directory then in your homedir. 
Create a ~/.config/containers/storage.conf With something like:
``` 
  graphroot = "/var/tmp/dwalsh/containers/storage"
```

**Comment by ashley-cui** [comment]
Closed via #1661
